### ISSUE TITLE
Ontologise the relations used in a collection

### ISSUE THREAD

**Issue by jesper-friis** [SINTEF/dlite#174]
Update the builtin relations in collections. Suggested changes

    _isA --> rdf:type (or maybe simply "a")
    _has-uuid --> dm:hasUUID
    _has-meta --> dm:hasMeta

Is it better to skip the namespace "dm" and just write `hasUUID` and `hasMeta`?

**Comment by francescalb** [comment]
Isn't putting definining dm as prefix enough?

**Comment by jesper-friis** [comment]
> Isn't putting definining dm as prefix enough?

Yes, we can do that when serialising a collection in turtle or other formats supporting prefixes. For the internal representation within the collection I think it is fine that dm is implicitly defined, since it is only used for a few hardcoded IRIs: dm:instanceOf, dm:hasUUID, dm:hasMeta, dm:Instance, dm:hasHash.

**Comment by jesper-friis** [comment]
> Personally I would prefer to use a better name than hasMeta. What is the main reason for not just using subClassOf and the inverse superClassOf instead of is-a and has-meta?

Good point. If we serialise both entities and instances as individuals as suggested in https://github.com/emmo-repo/datamodel/pull/10, we can use rdf:type instead of is-a. And the has-meta should be replaced with dm:instanceOf. Then we end up with the following:

    (<label>, rdf:type, dm:Instance)  # former _is-a
    (<label>, dm:hasUUID, <uuid>)     # former _has-uuid
    (<label>, dm:instanceOf, <uri>)   # former _has-meta
    (<label>, dm:hasHash, <hash>)     # for transactions

**Comment by francescalb** [comment]
Seems right for me.

**Comment by jesper-friis** [comment]
The only possible issue is that the idea with the label is that it is local to the collection, hence allowing simple labels like "energy". But that makes the relations invalid RDF, since the subject should be a [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#syntax).

One solution could be to keep it like this - to not enforce the relations to not be valid RDF. If the user needs that, he/she should  provide valid URIs as labels. Thoughts?

**Comment by jesper-friis** [comment]
Yes, I fully agree that we want to keep the collection simple and "user-friendly". I think that the collection backend in tripper might be the correct place to do the conversion between the simple `label` and the qualified `<namespace>/collection-id>#label` representations.

However, such conversions are not trivial, since we want reproduce the exact content of the triplestore after a conversion to and back from the collection representation. A further complication is that objects can either be IRIs or literals, so we need a way to distinguish between the two cases.

The only solution I see for that is to add support for prefixes in collections combined with some conventions. A possible set of conventions for labels and data objects could be:

* string with no colon (ex: "force") <--> string literal data object 
* string starting with a colon (ex: ":force") <--> an IRI with default prefix ("http://onto-ns.com/meta/0.1/Collection/<collection-id>#force")
* string of the form `<prefix>:<name>` <--> an IRI with given prefix

These conventions some drawbacks, like:
* it is not possible to represent data objects containing a colon, unless we introduce escaping
* we would have to prepend a colon to all subjects

**Comment by jesper-friis** [comment]
**Update:**
Instead of inventing a new syntax for literals, I think it may be easier and cleaner to move the type of a literal object to a fourth field in the internal dlite representation of a triple. If the type is NULL, the object is an IRI. Otherwise the object is a literal with the given type. An unqualified literal could be represented by an empty string (or a special pointer value, like `(void *)1`).

For IRIs, we can keep conventions 2-4 in the above comment. This would result in the following internal representation:

    ({label}, rdf:type, dm:Instance, NULL)           # former _is-a
    ({label}, dm:hasUUID, {uuid}, "")                # former _has-uuid
    ({label}, dm:instanceOf, {uri}, NULL)            # former _has-meta
    ({label}, dm:hasParent, {uuid}, "")              # for transactions
    ({label}, dm:hasHash, {hash}, "")                # for transactions

This will be easy to work with in C and convert to standard RDF representations via tripper or redland.

Since valid pointers must be word-aligned (a multiple of 4 for 32 bit systems and a multiple of 8 for 64 bit systems), there is plenty of space to encode common standard literal types in the pointer to the fourth field for very fast lookup using `switch()`.
